db/commitlog/commitlog.cc

```diff
@@ -1698,13 +1698,16 @@ future<> db::commitlog::segment_manager::shutdown() {
         // Wait for all pending requests to finish. Need to sync first because segments that are
         // alive may be holding semaphore permits.
         auto block_new_requests = get_units(_request_controller, max_request_controller_units());
-        return sync_all_segments().then([this, block_new_requests = std::move(block_new_requests)] () mutable {
+        return sync_all_segments().then_wrapped([this, block_new_requests = std::move(block_new_requests)] (future<> f) mutable {
+            if (f.failed()) {
+                clogger.error("Syncing all segments failed during shutdown: {}. Continuing with shutdown anyway...", f.get_exception());
+            }
```

Comment:
How can we fail here? What error?

I/O errors should be caught by the commitlog I/O error handler. We can't perform a clean shutdown, since we may lose data.

If it's a non-I/O error, then what is it?

Reply:
Hmm, according to https://github.com/scylladb/scylla/issues/8577#issue-874223808
> At 2021-04-25 02:22:45.517 an Enospc nemesis started on node 4 (13.53.173.30 | 10.0.1.229). The nemesis deliberately fills the target node's machine, clean it and afterwards restarts the scylla service.

Reply:
Yes, but what was the error?

Reply:
sync_all_segments ends up calling `segment::sync -> segment::cycle` which reports `_file.dma_write` errors to the log.
If this is the case, we should see https://github.com/scylladb/scylla/blob/4c351ff26009ece329231560cbcb0b54a468da0a/db/commitlog/commitlog.cc#L846

Reply:
@avikivity the question is what should be done if we can't ensure a clean shutdown?
Should the node just drain and hang?

Reply:
If there is an I/O error, we should isolate the node like we usually do. Provided we do that, it's okay to swallow the error (since the bad node can't affect the cluster without maintenance intervention). But shutting down cleanly isn't good, if the operator doesn't look at the logs, nobody will notice the error.

If it's not an I/O error, it should be fixed.

Reply:
@elcallio why don't we use `make_checked_file` in `db::commitlog::segment_manager::allocate_segment_ex`?
https://github.com/scylladb/scylla/blob/4c351ff26009ece329231560cbcb0b54a468da0a/db/commitlog/commitlog.cc#L1458

Reply:
I'm gonna go out on a limb and say because the (original) code was written long before check_file existed.

Reply:
Maybe we should use make_checked_file() then?

Reply:
@avikivity: this change now uses `make_checked_file` wrapping the opened file in `allocate_segment_ex`